src/include/codegen/query_cache.h

```diff
@@ -1,16 +1,46 @@
+//===----------------------------------------------------------------------===//
 //
-// Created by bechob on 4/11/17.
+//                         Peloton
 //
-
-#ifndef PELOTON_CACHE_H
-#define PELOTON_CACHE_H
-
-
-
-class cache {
-
+// value_peeker_proxy.h
+//
+// Identification: src/include/codegen/cache.h
+//
+// Copyright (c) 2015-17, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+#pragma once
+namespace peloton {
+namespace codegen {
+class QueryCache {
+public:
+  static QueryCache& Instance() {
+    static  QueryCache instance;
+    return instance;
+  }
+  bool FindPlan(const planner::AbstractPlan & key) {
+    auto it = cache.find(key);
+    if (it != cache.end())
+      return false;
+    return true;
+  }
+
+  Query GetQuery(const planenr::AbstractPlan & key) {
+    return cache.find(key)->second;
+  }
+
+  void InsertPlan(const planner::AbstractPlan & key, const Query & val) {
+    cache.insert(key, val);
+  }
+
+private:
+  struct ComparePlan {
+    bool operator()(const  planner::AbstractPlan & a, const planner::AbstractPlan & b) const {
+      return (codegen::PlanComparator::Compare(A, B) < 0);
+    }
+  };
+  std::map<planner::AbstractPlan, Query, ComparePlan> cache;
```

Comment:
A limit on the cache size would be nice to have since we would not have an infinite size of cache.  Having said this, I wonder what would be the best or a reasonable cache policy.